### ISSUE TITLE
Should maximal ranges be considered empty ranges? / Should ranges default to 0..-1?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#17605]
Typically, empty ranges have been represented in Chapel as the value `1..0` by default, on the basis that every index type with at least two values can represent it.  Another compelling empty range value to consider is `0..-1`, which has some modest attractive qualities:

* it can be thought of as being similar to the range `0..<0` which today amounts to `0..-1` and which feels appropriately default-y compared to, say, `1..<1` (arguably what we have today)
* it keeps the default low bound of the range `0`, as it would be in other "implicit indices" cases (and for some reason, making the low bound 0 seems to appeal more than having the high bound be 0).
* related: for index types with only a single value (like single-value enums), the low bound is representable in the index type (which, again, somehow appeals more than saying "the high bound is representable" today).

However, it also has a significant disadvantage, which is that unsigned integer types can't represent -1.

At the same time, we have significant problems or challenges with so-called "maximal" ranges (e.g., `0..max(uint(64)`), such as: (a) their size can't be represented using their index type, (b) it's challenging to create iterators that work correctly for them while also being efficient for more typical ranges.  These problems haven't been that problematic for us because maximal ranges don't come up very often in practice.

This combination of observations has led me to wonder whether maximal ranges (by some definition) ought to be considered an alternate form of empty range; and whether the default range value should be represented as `0..(-1):idxType`.

I don't think there's anything deeply pressing here, but I wanted to capture the notion for reference and possible discussion.

**Comment by bradcray** [comment]
https://github.com/chapel-lang/chapel/issues/16419 is vaguely related to this, if we consider any range with a high index of max(idxType) to be a maximal range.

**Comment by mppf** [comment]
> This combination of observations has led me to wonder whether maximal ranges (by some definition) ought to be considered an alternate form of empty range; and whether the default range value should be represented as 0..(-1):idxType.

I think it's important to point out that we don't need `0..max(int)` to become an empty range. It's unclear to me if you intend to propose that or if the maximal range considered here is only for unsigned types.

In any case, I do like the way that this proposal makes the default/empty range start from 0, which feels more natural in a 0-based language. I don't expect it has much impact on user code either way, but that's something to learn about with our tests if we decide this direction is worth considering.

**Comment by bradcray** [comment]
> I think it's important to point out that we don't need 0..max(int) to become an empty range.

From the perspective of being able to reason about its size, yes.  However, it's still a challenging range to iterate over efficiently.  I'm not sure whether you saw my most recent comment, but I think the challenging case for iteration is when `r.last + r.stride` overflows/underflows.  In any `low..max(int)` range, `max(int) + 1` does, which is what makes it a challenging case for iteration.

Alternatively, we could say that such ranges are not empty, but simply don't support certain operations, like iteration (which is more or less what we do today).  For me, the question then becomes, what is the line between maximal empty ranges and other maximal ranges?  E.g., `0:uint..<0:uint` is "obviously" empty by this proposal, but what about `low:uint..<0:uint`?

**Comment by mppf** [comment]
Right. I think it becomes related to whether or not the range element type is normally `int`. In particular I think it's important that `0:uint(8)..255:uint(8)` or `0:uint(8)..<0:uint(8)` can work but I'm not really worried about ranges near the maximum of 64-bit integers.

**Comment by bradcray** [comment]
For those following this issue on email, I've edited the OP to list a number of options that we could consider here w.r.t. language stabilization (including having the spec say that the bounds of a default-initialized empty range are not well-defined).
